pydrofoil/real.py

```diff
@@ -0,0 +1,264 @@
+                num = rbigint.fromint(-num)
+                den = rbigint.fromint(-den)
+        elif ((num > MININT and num < MAXINT) and (den > MININT and den < MAXINT)):
+            sign = num/abs(num)*den/abs(den)
+            num = rbigint.fromint(abs(num)*sign)
+            den = rbigint.fromint(abs(den))
+        else:
+            assert False, "input is out of range of INT"
+        # assert not den.int_eq(0), "denominator cannot be 0"
+        return Real(num, den)
+    
+    @staticmethod
+    def fromstr(str):
+        from rpython.rlib.rstring import strip_spaces
+        s = strip_spaces(str)
+        j = -1
+        lead_zero_count = 0
+        ZERO_COUNT_END = False
+        for i in range(0, len(s)):
+            if s[i].isdigit() and s[i] != "0":
+                ZERO_COUNT_END = True
+            if (not ZERO_COUNT_END) and s[i] == "0":
+                lead_zero_count += 1 
+            if not s[i].isdigit() and s[i] != "+" and s[i] != "-":
+                if i == 1 or (i == 2 and (s[0] == "+" or s[0] == "-")):
+                    for j in range(len(s)-1, i, -1):
+                        if (s[j] == "+" or s[j] == "-") and s[j-1] == "e":
+                            break
+                for k in range(i+1, len(s)):
+                    if (not ZERO_COUNT_END) and s[k] == "0":
+                        lead_zero_count += 1
+                    else:
+                        break
+                break
+        # return i, s[:i]+s[i+1:]
+        if j == -1 or (j == 2 and (s[0] != "+" and s[0] != "-")) or (j == 3 and (s[0] == "+" or s[0] == "-")):
+            if lead_zero_count == len(s)-1:
+                return Real(rbigint.fromint(0), rbigint.fromint(1))
+            num = rbigint.fromstr((s[:i] + s[i+1:])[lead_zero_count:] if (s[0] != "+" and s[0] != "-") else s[0]+(s[1:i] + s[i+1:])[lead_zero_count:]) if i < len(s)-1 else rbigint.fromstr(s)
+            dif = len(s)-1-i
+            if dif < 19:
+                den = rbigint.fromint(10**dif)
+            else:
+                den = den_of_fromstr(dif)
+            return Real(num, den)
+        else:
+            if s[j] == "+":
+                shift = int(s[j+1:])
+                if shift >= j-i-2:
+                    num = rbigint.fromstr(s[:i]+s[i+1:j-1]+"0"*(shift-(j-i-2)))
+                    den = rbigint.fromint(1)
+                    return Real(num, den)
+                else:
+                    return Real.fromstr(s[:i]+s[i+1:i+1+shift]+s[i]+s[i+1+shift:j-1])
+                # 1.222e+5
+                # 1.222222e+3
+            elif s[j] == "-":
+                shift = int(s[j+1:])
+                num = rbigint.fromstr(s[:i]+s[i+1:j-1])
+                dif = j-i-2+shift
+                if dif < 19:
+                    den = rbigint.fromint(10**dif)
+                else:
+                    den = den_of_fromstr(dif)
+                return Real(num, den)
+
+    # string to real without considering about input as x.xxxxe+xxx or x.xxxxe-xxxx
```

Comment:
I think we can go with the simpler version until we know for sure that more complicated cases are needed by sail. I checked which arguments the ARM model uses for fromstr, and it's only the following strings:

```
"0.0"
"0.5"
"1.0"
"2.0"
"3.0"
```

Reply:
both versions look really complicated though. it somehow feels like it should be much simpler to express this, using Python string methods. I sketched a bit the simple version, untested:

```python
s = strip_spaces(str)
sign = 1
if s.startswith('-'):
    s = s[1:]
    sign = -1
s = s.lstrip('0') # remove leading zeros
decimalpos = s.find(".")
if decimalpos == -1: # not found
    return Real(rbigint.fromstr(s).int_mul(sign), rbigint.fromint(1))
den = rbigint.fromint(10).int_pow(len(s) - decimalpos)
num = rbigint.fromint(s.replace('.', ''))
return Real(num.int_mul(sign), den)
```

Reply:
yes, I'm going to change it cause I'm not so sure what format we have for the input, so I just try to make it accepts more formats, and also for leading zeros, I checked that if we use rbigint.fromstr(), and set the base to 10, it can deal with the leading zeros automatically